File: app/src/main/java/com/tronprotocol/app/selfmod/ValidationResult.java
package com.tronprotocol.app.selfmod;

import java.util.ArrayList;
import java.util.List;

/**
 * Result of code modification validation
 */
public class ValidationResult {
    private boolean isValid;
    private final List<String> errors;
    private final List<String> warnings;
    
    public ValidationResult() {
        this.isValid = false;
        this.errors = new ArrayList<>();
        this.warnings = new ArrayList<>();
    }
    
    public void addError(String error) {
        errors.add(error);
        isValid = false;
    }
    
    public void addWarning(String warning) {
        warnings.add(warning);
    }
    
    public void setValid(boolean valid) {
        this.isValid = valid;
    }
    
    public boolean isValid() {
        return isValid && errors.isEmpty();
    }
    
    public List<String> getErrors() {
        return new ArrayList<>(errors);
    }
    
    public List<String> getWarnings() {
        return new ArrayList<>(warnings);
    }
    
    @Override
    public String toString() {
        return "ValidationResult{" +
                "valid=" + isValid +
                ", errors=" + errors.size() +
                ", warnings=" + warnings.size() +
                '}';
    }
}

File: app/src/main/java/com/tronprotocol/app/selfmod/ValidationResult.kt
package com.tronprotocol.app.selfmod

/**
 * Result of staged code modification validation and gating.
 */
class ValidationResult {

    enum class Stage {
        PROPOSAL,
        STATIC_CHECKS,
        SANDBOX_RUN,
        CANARY_ROLLOUT,
        FULL_ROLLOUT,
        ROLLED_BACK
    }

    data class GateResult(
        val gateName: String,
        val passed: Boolean,
        val details: String,
        val timestamp: Long = System.currentTimeMillis()
    )

    private var valid: Boolean = false
    private var stage: Stage = Stage.PROPOSAL
    private val errors = mutableListOf<String>()
    private val warnings = mutableListOf<String>()
    private val gates = mutableListOf<GateResult>()

    fun addError(error: String) { errors.add(error); valid = false }
    fun addWarning(warning: String) { warnings.add(warning) }
    fun setValid(valid: Boolean) { this.valid = valid }
    fun setStage(stage: Stage) { this.stage = stage }

    fun addGateResult(gateName: String, passed: Boolean, details: String) {
        gates.add(GateResult(gateName = gateName, passed = passed, details = details))
        if (!passed) valid = false
    }

    fun isValid(): Boolean = valid && errors.isEmpty()
    fun getStage(): Stage = stage
    fun getErrors(): List<String> = ArrayList(errors)
    fun getWarnings(): List<String> = ArrayList(warnings)
    fun getGateResults(): List<GateResult> = ArrayList(gates)

    override fun toString(): String {
        return "ValidationResult{" +
                "valid=$valid" +
                ", stage=$stage" +
                ", gates=${gates.size}" +
                ", errors=${errors.size}" +
                ", warnings=${warnings.size}" +
                "}"
    }
}
